pkg/preparer/podprocess/reporter.go

```diff
@@ -0,0 +1,313 @@
+		}
+	}
+
+	workspaceFilePath := r.workspaceFilePath()
+	// Check that the workspace file either:
+	// 1) exists and is readable and writable by root
+	// OR
+	// 2) does not exist and is successfully created.
+	info, err := os.Stat(workspaceFilePath)
+	if os.IsNotExist(err) {
+		file, err := os.Create(workspaceFilePath)
+		if err != nil {
+			return util.Errorf("Could not create workspace file at %s: %s", workspaceFilePath, err)
+		}
+
+		err = file.Chmod(0600)
+		if err != nil {
+			return util.Errorf("Could not chmod workspace file at %s: %s", workspaceFilePath, err)
+		}
+
+		// Initialize the file with an id of 0
+		err = r.writeLastID(0)
+		if err != nil {
+			return err
+		}
+	} else if err != nil {
+		return util.Errorf("Error running stat() on %s: %s", workspaceFilePath, err)
+	} else {
+		if info.Mode()&0600 != 0600 {
+			return util.Errorf("Workspace file %s exists but is not readable and writable by root: perms %s", workspaceFilePath, info.Mode())
+		}
+	}
+
+	return nil
+}
+
+func (r *Reporter) openAndMaybeMigrateDatabase() (*sql.DB, error) {
+	db, err := sql.Open("sqlite3", r.sqliteDBPath)
+	if err != nil {
+		return nil, util.Errorf("Could not open sqlite database: %s", err)
+	}
+
+	sqlStmt := `
+	create table if not exists finishes (
+	    id integer not null primary key,
+	    date datetime default current_timestamp,
+	    pod_id text,
+	    pod_unique_key text,
+	    launchable_id text,
+	    entry_point text,
+	    exit_code integer,
+	    exit_status integer
+	);
+	`
+	_, err = db.Exec(sqlStmt)
+	if err != nil {
+		_ = db.Close()
+		return nil, util.Errorf("Could not run create table statement: %s", err)
+	}
+	return db, nil
+
+}
+
+func (r *Reporter) publishProcessExits(db *sql.DB, quitCh <-chan struct{}) {
+	defer db.Close()
+	timer := time.NewTimer(0)
+	for {
+		select {
+		case <-quitCh:
+			return
+		case <-timer.C:
+			r.reportLatestExits(db)
+			timer.Reset(r.pollInterval)
+		}
+	}
+}
+
+func (r *Reporter) reportLatestExits(db *sql.DB) {
+	lastID, err := r.getLastID()
+	if err != nil {
+		r.logger.WithError(err).Errorln("Could not read last ID from process exit sqlite db")
+		return
+	}
+
+	rows, err := db.Query(`
+	    SELECT id, date, pod_id, pod_unique_key, launchable_id, entry_point, exit_code, exit_status
+	    FROM finishes
+	    WHERE id > ?
+	    `, lastID)
+	if err != nil {
+		r.logger.WithError(err).Errorln("Could not query for latest process exits")
+		return
+	}
+	defer rows.Close()
+
+	var id int64
+	var date time.Time
+	var podID, podUniqueKey, launchableID, entryPoint string
+	var exitCode, exitStatus int
+
+	defer func() {
+		err = r.writeLastID(lastID)
+		if err != nil {
+			r.logger.WithError(err).Errorln("Could not write last ID to workspace file, duplicate messages may be reported")
+		}
+	}()
+	for rows.Next() {
+		err = rows.Scan(&id, &date, &podID, &podUniqueKey, &launchableID, &entryPoint, &exitCode, &exitStatus)
```

Comment:
I'm curious for the motivation here? the function doesn't do much else (the only other thing is `r.getLastID()`) and I don't anticipate other places to need to do the scan